### PR TITLE
Enable registration of add-on providers

### PR DESCRIPTION
**Pull request by ewinston**

### CODE REVIEW

File: qiskit/wrapper/_wrapper.py [hunk 1]
```diff
@@ -19,13 +19,16 @@
...
-             hub=None, group=None, project=None, proxies=None, verify=True,
-             provider_name=None):
+def register(*args, provider_class=IBMQProvider, provider_name='ibmq', **kwargs):
     """
     Authenticate against an online backend provider.
 
     Args:
+        args (tuple): positional arguments passed to provider class initialization
+        provider_class (BaseProvider): provider class
+        provider_name (str): name for provider
```
Comment (ajavadia) [inline qiskit/wrapper/_wrapper.py]: I would explain this a bit more, e.g. "optional user-assigned name for the provider"

File: qiskit/wrapper/_wrapper.py [hunk 2]
```diff
@@ -36,17 +39,9 @@ def register(token, url='https://quantumexperience.ng.bluemix.net/api',
             proxies (dict): Proxy configuration for the API, as a dict with
                 'urls' and credential keys.
             verify (bool): If False, ignores SSL certificates errors.
```
Comment (ajavadia) [inline qiskit/wrapper/_wrapper.py]: The old options should be deleted from docstring.

**Comment by ajavadia** [comment]
I rebased against master. Removed provider names introduced in #584. Need to make some modifications to tests, but basically done. It can be used like this:

```
from qiskit import *
available_backends()  # shows local simulators

import Qconfig
register(Qconfig.APItoken)
available_backends()  # show remote devices

from qiskit_addon_sympy import SympyProvider
register(provider_class=SympyProvider)
available_backends()  # show sympy simulators
```

**Comment by ajavadia** [comment]
All but 2 tests pass. @diego-plan9, can you please take a look? I tried to keep the original intent of your wrapper tests but was not sure about 2 cases:

1. `test_register_backend_name_conflicts`: i think the tests you have written here make a lot of sense. I tried to do this using two accounts (perhaps private/public), and show that the second provider does get registered (but the `available_backends()` belong to the first one). However I don't know how to do this registration using 2 tokens. Can we mock or something?

2. `test_register_bad_credentials`: do we want a QISKitError raised when the provider fails to initialize for whatever reason? If so, this should presumably go as a try-except in `wrapper.register()`. But I don't know how to avoid catching a broad except, because the failure could be for many reasons.

**Comment by atilag** [comment]
Hi Ali,
>However I don't know how to do this registration using 2 tokens. Can we mock or something?

What's happening here is that the URL is pointing to the public backends, so no matter what token are you passing, you will always get the same backend names, so they are registered but can't be accessed.

> do we want a QISKitError raised when the provider fails to initialize for whatever reason? 

I think It's safe to wrap the error raised from the provider registration into a QISKitError for now.

**Pull request merged**